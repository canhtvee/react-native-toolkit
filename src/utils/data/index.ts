export * from './Fetching';
export * from './Session';
