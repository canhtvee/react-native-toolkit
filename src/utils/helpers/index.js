export * from './Convert';
export * from './ResetFunction';
export * from './Validate';
