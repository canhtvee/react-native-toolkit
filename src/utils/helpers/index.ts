export * from './Convert';
export * from './ResetFunction';
export * from './Mocking';
export * from './Constants';
