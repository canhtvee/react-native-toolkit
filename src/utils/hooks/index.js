export * from './useFocusRefetch';
export * from './useRerender';
