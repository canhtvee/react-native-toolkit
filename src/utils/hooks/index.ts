export * from './useFocusRefetch';
export * from './useRerender';
export * from './useUnmountEffect';
export * from './useSetTimeout';
