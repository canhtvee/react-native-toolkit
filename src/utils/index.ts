export * from './data';
export * from './helpers';
export * from './hooks';
export * from './modules';
export * from './resources';
export * from './types';
