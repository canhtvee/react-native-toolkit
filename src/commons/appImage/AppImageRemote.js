import React, {useEffect, useState} from 'react';
import {View} from 'react-native';
import FastImage from 'react-native-fast-image';

import {Constants, useAppContext} from '../../utils';

import {AppViewLoading} from '../appViewLoading';
import {AppIcon} from '../appIcon';

import {styles} from './styles';

const _getResizeMode = resizeMode => {
  switch (resizeMode) {
    case 'contain':
      return FastImage.resizeMode.contain;
    default:
      return FastImage.resizeMode.cover;
  }
};

/**image source validation **/
const _isValidSource = source => {
  if (!source || !source?.uri) {
    return false;
  }
  // if (url.substring(0, 4) !== 'http') {
  //   return false;
  // }
  return source.uri.match(/\.(jpeg|jpg|gif|png|JPEG|JPG|PNG)$/) !== null;
};

export function AppImageRemote({
  resizeMode,
  imageStyle,
  imageContainerStyle,
  source,
  placeholder,
  spinnerColor,
  spinnerSize,
  isLoading,
  onSuccess,
  onError,
  ...imageProps
}) {
  const {Colors} = useAppContext();
  const [imageStatus, setImageStatus] = useState(() =>
    _isValidSource(source) || isLoading
      ? Constants.STATUS_LOADING
      : Constants.STATUS_ERROR,
  );

  useEffect(() => {
    if (isLoading) {
      setImageStatus(Constants.STATUS_LOADING);
    }
  }, [isLoading]);

  useEffect(() => {
    setImageStatus(
      _isValidSource(source)
        ? Constants.STATUS_LOADING
        : Constants.STATUS_ERROR,
    );
  }, [source]);

  if (imageStatus === Constants.STATUS_ERROR) {
    return (
      <View style={imageContainerStyle}>
        {placeholder ? (
          placeholder
        ) : (
          <AppIcon name={'alert-triangle'} size={24} />
        )}
      </View>
    );
  }

  return (
    <View style={imageContainerStyle}>
      <FastImage
        style={[styles.image, imageStyle]}
        source={source}
        resizeMode={_getResizeMode(resizeMode)}
        onLoad={() => {
          setImageStatus(Constants.STATUS_SUCCESSFUL);
          onSuccess && onSuccess();
        }}
        onError={() => {
          setImageStatus(Constants.STATUS_ERROR);
          onError && onError();
        }}
        {...imageProps}>
        {imageStatus === Constants.STATUS_LOADING && (
          <AppViewLoading
            spinnerSize={spinnerSize}
            spinnerColor={spinnerColor || Colors.onBackground}
            containerStyle={[
              {backgroundColor: Colors.withAlpha(Colors.background, '20%')},
              imageStyle,
            ]}
          />
        )}
      </FastImage>
    </View>
  );
}
