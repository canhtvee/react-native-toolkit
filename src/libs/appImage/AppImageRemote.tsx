import React, {useEffect, useState} from 'react';
import {View} from 'react-native';
import FastImage, {
  FastImageProps,
  Priority,
  ResizeMode,
} from 'react-native-fast-image';

import {addAlpha, ResourceStatusType, useAppContext} from '../../utils';

import {AppViewLoading} from '../appViewLoading';
import {AppIcon} from '../appIcon';

export type AppImageRemoteSourceType = {
  uri?: string;
  headers?: {
    [key: string]: string;
  };
  priority?: Priority;
  cache?: 'immutable' | 'web' | 'cacheOnly';
};

export interface AppImageRemoteProps
  extends Omit<FastImageProps, 'source' | 'onLoad' | 'onError'> {
  source: AppImageRemoteSourceType;
  spinnerSize?: number;
  spinnerColor?: string;
  placeholder?: JSX.Element;
  isLoading?: boolean;
  onSuccess?: () => void;
  onError?: () => void;
}

const getResizeMode = (resizeMode?: ResizeMode) => {
  switch (resizeMode) {
    case 'contain':
      return FastImage.resizeMode.contain;
    default:
      return FastImage.resizeMode.cover;
  }
};
/**image url **/
const isValidUrl = (url: string) => {
  if (!url) {
    return false;
  }
  // if (url.substring(0, 4) !== 'http') {
  //   return false;
  // }
  return url.match(/\.(jpeg|jpg|gif|png|JPEG|JPG|PNG)$/) !== null;
};

export function AppImageRemote({
  resizeMode,
  style,
  source,
  placeholder,
  spinnerColor,
  spinnerSize,
  isLoading,
  onSuccess,
  onError,
  ...imageProps
}: AppImageRemoteProps) {
  const {Colors} = useAppContext();
  const [imageStatus, setImageStatus] = useState<ResourceStatusType>(() =>
    source && source.uri && isValidUrl(source.uri) ? 'loading' : 'error',
  );

  useEffect(() => {
    setImageStatus(
      source && source.uri && isValidUrl(source.uri) ? 'loading' : 'error',
    );
  }, [source]);

  if (imageStatus === 'loading' || imageStatus === 'successful' || isLoading) {
    return (
      <FastImage
        style={[
          {
            width: 86,
            height: 86,
            justifyContent: 'center',
            alignItems: 'center',
          },
          style,
        ]}
        source={source}
        resizeMode={getResizeMode(resizeMode)}
        onLoad={() => {
          setImageStatus('successful');
          onSuccess && onSuccess();
        }}
        onError={() => {
          setImageStatus('error');
          onError && onError();
        }}
        {...imageProps}>
        {(imageStatus === 'loading' || isLoading) && (
          <AppViewLoading
            spinnerSize={spinnerSize}
            spinnerColor={spinnerColor}
            containerStyle={[
              {backgroundColor: addAlpha(Colors.surface, 0.54)},
              style,
            ]}
          />
        )}
      </FastImage>
    );
  }
  return (
    <View
      style={[
        {
          justifyContent: 'center',
          alignItems: 'center',
          width: 86,
          height: 86,
        },
        style,
      ]}>
      {placeholder ? (
        placeholder
      ) : (
        <AppIcon name={{feather: 'alert-triangle'}} size={24} />
      )}
    </View>
  );
}
