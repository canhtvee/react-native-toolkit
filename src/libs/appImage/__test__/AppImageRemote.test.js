import React from 'react';
import {View} from 'react-native';

import {render, cleanup} from '@testing-library/react-native';

import {AppImageRemote} from '../AppImageRemote';
import {AppContextProvider} from '../../../utils';

/**
 * Run test command
 * yarn test -u -t="AppImageRemote"
 */

afterAll(cleanup);

describe('AppImageRemote', () => {
  test('should render correctly', () => {
    const tree = render(
      <AppContextProvider>
        <AppImageRemote
          source={{
            uri: 'https://i.pinimg.com/564x/7c/93/2e/7c932e5fc8a86804082cd22203455f8f.jpg',
          }}
          style={{
            backgroundColor: 'red',
            margin: 20,
          }}
        />
      </AppContextProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
});
