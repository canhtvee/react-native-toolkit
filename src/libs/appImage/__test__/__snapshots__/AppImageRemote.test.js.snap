// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppImageRemote should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Array [
        Object {
          "alignItems": "center",
          "height": 86,
          "justifyContent": "center",
          "width": 86,
        },
        Object {
          "backgroundColor": "red",
          "margin": 20,
        },
      ],
    ]
  }
>
  <FastImageView
    onFastImageError={[Function]}
    onFastImageLoad={[Function]}
    resizeMode="cover"
    source={
      Object {
        "uri": "https://i.pinimg.com/564x/7c/93/2e/7c932e5fc8a86804082cd22203455f8f.jpg",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 6,
          "paddingRight": 6,
        },
        Array [
          Object {
            "backgroundColor": "#FFFFFF8A",
          },
          Object {
            "backgroundColor": "red",
            "margin": 20,
          },
        ],
      ]
    }
  >
    <RNSpinkit
      color={4278190080}
      size={24}
      style={
        Array [
          Object {
            "height": 24,
            "width": 24,
          },
          undefined,
        ]
      }
      type="Circle"
    />
  </View>
</View>
`;
